From 19d0874456b7922127ef25d014aea661546721ae Mon Sep 17 00:00:00 2001
From: Li Jun <jun.li@nxp.com>
Date: Tue, 19 Dec 2017 19:07:44 +0800
Subject: [PATCH 3116/5242] MLK-17242-1 staging: typec: tcpci: add sink
 setting for sink-disable

commit  8c3a6d4ba683e7ac9ddd56fcbc267758e6c574e5 from
https://source.codeaurora.org/external/imx/linux-imx.git

Adding fixed sink power settings for sink-disable case, which
is only for PD protocol talk to know the cc orientation if connects
to a PD capable host, HW doesn't really sink any power in this case.

Tested-by: Peter Chen <peter.chen@nxp.com>
Signed-off-by: Li Jun <jun.li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/staging/typec/tcpci.c |   40 +++++++++++++++++++++++++++++++++-------
 1 file changed, 33 insertions(+), 7 deletions(-)

diff --git a/drivers/staging/typec/tcpci.c b/drivers/staging/typec/tcpci.c
index 37e3601..c84dd00 100644
--- a/drivers/staging/typec/tcpci.c
+++ b/drivers/staging/typec/tcpci.c
@@ -717,6 +717,39 @@ static int tcpci_parse_config(struct tcpci *tcpci)
 		return -EINVAL;
 	}
 
+	/*
+	 * In case DRP only for data role, power role is source only
+	 * we can use this property to disable power sink.
+	 */
+	if (device_property_read_bool(tcpci->dev, "sink-disable")) {
+		tcpci->sink_disable = true;
+
+		/* Provide a sink PDO to setup a PD session */
+		tcfg->nr_snk_pdo = 1;
+		tcfg->snk_pdo = devm_kzalloc(tcpci->dev,
+				sizeof(*tcfg->snk_pdo), GFP_KERNEL);
+		if (!tcfg->snk_pdo)
+			return -ENOMEM;
+
+		/*
+		 * Sink PDO setting:
+		 * - Voltage in 50mV units: 5V
+		 * - Operational Current in 10mA units: 100mA
+		 */
+		*tcfg->snk_pdo = PDO_FIXED(5000,
+					100,
+					PDO_FIXED_DUAL_ROLE |
+					PDO_FIXED_EXTPOWER |
+					PDO_FIXED_USB_COMM |
+					PDO_FIXED_DATA_SWAP);
+		tcfg->max_snk_mv = 5000;
+		tcfg->max_snk_ma = 2000;
+		tcfg->max_snk_mw = 10000;
+		tcfg->operating_snk_mw = 500;
+
+		return 0;
+	}
+
 	/* Check the num of snk pdo */
 	tcfg->nr_snk_pdo = device_property_read_u32_array(tcpci->dev,
 						"snk-pdos", NULL, 0);
@@ -751,13 +784,6 @@ static int tcpci_parse_config(struct tcpci *tcpci)
 						&tcfg->operating_snk_mw))
 		goto snk_setting_wrong;
 
-	/*
-	 * In case DRP only for data role, power role is source only
-	 * we can use this property to disable power sink.
-	 */
-	if (device_property_read_bool(tcpci->dev, "sink-disable"))
-		tcpci->sink_disable = true;
-
 	return 0;
 
 snk_setting_wrong:
-- 
1.7.9.5

